### DISCUSSION TITLE
BaseModel and Events

### DISCUSSION THREAD

**Discussion by veera-visw** [pybamm-team/PyBaMM#3882] [Q&A] [answered]
I wish to add events to my model which is part of the BaseModel class. I tried the following:

model.events = [
    pybamm.Event(
        "Minimum voltage",
        BV-low_cutoff,
    ),
        pybamm.Event(
            "Minimum positive particle surface stoichiometry",
            pybamm.min(c_surf/c_max) - 0.01,
        ),

]

here: BV is the Battery voltage and low_cutoff is the pybamm.Parameter () for lower cut off. 

When I run this code, I get the following error:
TypeError: 
                Cannot convert symbol of type '<class 'pybamm.expression_tree.parameter.Parameter'>' to CasADi. Symbols must all be
                'linear algebra' at this stage.

Any suggestions to fix this ? 

THanks

**Comment by martinjrobins** [comment]
Hi @veera-visw, you are getting this error as you still have `pybamm.Parameter` nodes in your model, these should have been replaced by actual values via the `pybamm.ParameterValues` class. This usually happens automatically using the `pybamm.Simulation` class. Are you able to share your full code, then I can give you more specific advice.

**Comment by veera-visw** [comment]
@martinjrobins  Thanks for the reply. I assumed the same thing ! My code as such is quite dirty at the moment so let me clean it up and share it with you ASAP.

Meanwhile, the general structure of my code looks something like this (not a fully working code, just a skeleton)

model=pybamm.BaseModel()
BV=pybamm.parameter("Voltage [V]")
low_cutoff=pybamm.Parameter("Lower voltage cut-off [V]")
// equations based on DFN model

model.variables={
"Voltage [V"]: "pybamm.boundary_value(phi,"right")
}
param=pybamm.ParameterValues(
{
"Lower voltage cut-off [V]" : 0.3,
}
model.events = [
pybamm.Event(
"Minimum voltage",
BV-low_cutoff,
)]
solution=solver.solve(model,t_eval)


I dont think I am using the pybamm.Simulation() class rather directly using the solver.solve class (similar to the half cell example)

**Reply by martinjrobins** [reply, marked as answer]
so you are not inserting the values in `param` into your model. The easiest way to do this is use the `pybamm.Simulation` class like so:

```python
sim = pybamm.Simulation(model, parameter_values=param)
sol = sim.solve(t_eval)
```

without the Simulation class, you can use the ParameterValues class directly:

```python
param.process_model(model)
param.process_geometry(geometry)
```

**Comment by veera-visw** [comment]
@martinjrobins  Good catch!! I went through my code again and I had the model.events() after the process_model(model). Silly mistake. but seems to be fixed now :)

**Reply by martinjrobins** [reply]
great to hear its working @veera-visw  :) for organisation purposes, can you please mark this discussion as answered? Thanks!
